src/online/index.js

```diff
@@ -1,295 +1,267 @@
-        throw new Error('Parse Error ' + error.code + ': ' + error.message)
-      }
-    })
-  }
-
-  function unwrapUser(parseUser) {
-    if (!parseUser) return null
-    return {
-      username: parseUser.get('username'),
-      email:    parseUser.get('email'),
-    }
-  }
+  const service = new OnlineService()
 
-  function toObject(物) {
-    return Object.assign({ }, 物.attributes, { id: 物.id })
-  }
+  const user口 = new Bacon.Bus()
+  const user川 = (
+    user口
+    // https://github.com/baconjs/bacon.js/issues/536
+    .toProperty(null)
+    .map(user => user || service.getCurrentUser())
+  )
 
-  function signUp({ username, password, email }) {
-    invariant(typeof username === 'string', 'username must be a string')
-    invariant(typeof password === 'string', 'password must be a string')
-    invariant(typeof email === 'string',    'email must be a string')
-    return (
-      wrapPromise(Parse.User.signUp(username, password, { email }))
-      .tap(user => user口.push(user))
-    )
+  function signUp(options) {
+    return service.signUp(options).tap(user => user口.push(user))
   }
 
-  function logIn({ username, password }) {
-    invariant(typeof username === 'string', 'username must be a string')
-    invariant(typeof password === 'string', 'password must be a string')
-    return (
-      wrapPromise(Parse.User.logIn(username, password))
-      .tap(user => user口.push(user))
-    )
+  function logIn(options) {
+    return service.logIn(options).tap(user => user口.push(user))
   }
 
   function logOut() {
-    return (
-      wrapPromise(Parse.User.logOut()).then(() => {})
-      .tap(() => user口.push(null))
-    )
+    return service.logOut().tap(() => user口.push(null))
   }
 
   function submitScore(info) {
-    return (
-      wrapPromise(Parse.Cloud.run('submitScore', info))
-      .then(({ data, meta }) => {
-        return {
-          data: toObject(data),
-          meta: meta
-        }
-      })
-    )
+    return service.submitScore(info)
   }
 
-  function parseRetrieveRecord(descriptor) {
-    let { md5, playMode } = descriptor
-    let query = new Parse.Query('GameScore')
-    query.equalTo('md5',      md5)
-    query.equalTo('playMode', playMode)
-    query.equalTo('user',     Parse.User.current())
-    return wrapPromise(query.first()).tap(record => {
-      cache.set(descriptor.recordCacheKey(), record)
-    })
+  function getScoreboard(descriptor) {
+    return service.retrieveScoreboard(descriptor)
   }
 
-  function parseRetrieveRank(descriptor, gameScore) {
-    let { md5, playMode } = descriptor
-    let countQuery = new Parse.Query('GameScore')
-    countQuery.equalTo('md5',       md5)
-    countQuery.equalTo('playMode',  playMode)
-    countQuery.greaterThan('score', gameScore.get('score'))
-    return wrapPromise(countQuery.count())
-  }
+  const putRecord口  = new Bacon.Bus()
+  const wantRecord口 = new Bacon.Bus()
+  const seen口       = new Bacon.Bus()
+  const records     = new DataStore(putRecord口)
```

Comment:
- Make sure that `records` is cleared when log out.
- Make sure that `records` is reloaded when log in.